src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Publish/AvoidAssemblyLocationInSingleFile.cs

```diff
@@ -0,0 +1,158 @@
+    {
+        public const string IL3000 = nameof(IL3000);
+        public const string IL3001 = nameof(IL3001);
+
+        internal static DiagnosticDescriptor LocationRule = DiagnosticDescriptorHelper.Create(
+            IL3000,
+            new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.AvoidAssemblyLocationInSingleFileTitle),
+                MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources)),
+            new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.AvoidAssemblyLocationInSingleFileMessage),
+                MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources)),
+            DiagnosticCategory.Publish,
+            RuleLevel.BuildWarning,
+            description: null,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        internal static DiagnosticDescriptor GetFilesRule = DiagnosticDescriptorHelper.Create(
+            IL3001,
+            new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.AvoidAssemblyGetFilesInSingleFile),
+                MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources)),
+            new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.AvoidAssemblyGetFilesInSingleFile),
+                MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources)),
+            DiagnosticCategory.Publish,
+            RuleLevel.BuildWarning,
+            description: null,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(LocationRule, GetFilesRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var compilation = context.Compilation;
+                var isSingleFilePublish = context.Options.GetMSBuildPropertyValue(
+                    MSBuildPropertyOptionNames.PublishSingleFile, compilation, context.CancellationToken);
+                if (!string.Equals(isSingleFilePublish?.Trim(), "true", StringComparison.OrdinalIgnoreCase))
+                {
+                    return;
+                }
+                var includesAllContent = context.Options.GetMSBuildPropertyValue(
+                    MSBuildPropertyOptionNames.IncludeAllContentForSelfExtract, compilation, context.CancellationToken);
+                if (string.Equals(includesAllContent?.Trim(), "true", StringComparison.OrdinalIgnoreCase))
+                {
+                    return;
+                }
+
+                var properties = new List<IPropertySymbol>();
```

Comment:
Will this data structure be thread-safe given we will get concurrent callbacks?

Reply:
Might be best to use an ImmutableHashSet/ImmutableArray for these.

Reply:
If you decide to use ImmutableHashSet, then you can use this extension method: https://github.com/dotnet/roslyn-analyzers/blob/194d1443c4f5901afd4142587f88c35b06b37654/src/Utilities/Compiler/Extensions/ImmutableHashSetExtensions.cs#L95

Reply:
This should be safe given that threading model. ImmutableHashSet is probably a bad idea for lists of < 8 elements. I can move to ImmutableArray instead.